### DISCUSSION TITLE
Help running tests in a browser in GitHub Actions

### DISCUSSION THREAD

**Discussion by ericoporto** [emscripten-core/emscripten#21578] [Q&A] [answered]
Hey, I have a wasm/js/HTML thing built. I would like to run it on a browser in GitHub Actions CI and print the console.log output to the CI log and if possible also grab a file that is generated in the MEMFS/IDBFS.

Any ideas on this are welcome. Asking here because I imagine more people have the need to test their code in CI pipelines that are also here.

**Comment by kripken** [comment, marked as answer]
In general you can run browsers in headless mode on CI. Here is where Emscripten itself does that:

https://github.com/emscripten-core/emscripten/blob/ecdca7ba001306e4361584bbfcd873945143434f/.circleci/config.yml#L311

Note the use of "new" headless mode in Chrome there, which works better in our experience.

Running with `emrun` should log out stdout/stderr to the console. I don't think it captures `console.log` as well but that should be easy to hook up too.

**Reply by ericoporto** [reply]
Nice, that is really useful information! Thank you!!

**Reply by sbc100** [reply]
You can also use the selenium webdriver to drive your tests.  Googling "selenium github actions" gives a bunch of resuts.

**Reply by ericoporto** [reply]
Thank you @sbc100 !!! Will check that too!!
